### PR TITLE
Gimesketvirtadien's  time requested changes

### PR DESCRIPTION
**Pull request by KjellKod**
Inherited changes from: gimesketvirtadieni's https://github.com/KjellKod/g3log/pull/90

- added support for OSX (get_timespec for now, on Linux)
- fixed formatting
- corrected default logging header
- started with simplified logic (skip c-arrays)... it can be renamed better though

** not yet unit tests which is sad. It'll be added! this was exploratory coding more than TDD
** not yet Windows support.

### CODE REVIEW

**Comment by KjellKod** [comment]
@gimesketvirtadieni, travis fails but I think that has to do with the compiler being too old. I will investigate that later.

File: src/time.cpp [hunk 1]
```diff
@@ -2,17 +2,53 @@
...
+   const size_t kNsDigitsToCut = 1; // all digits left : divide by 1
+   const size_t kNsDefaultDigits = 0;
+   const size_t kDigitIdentifier = 1;
+   const size_t kNoDigitIdentifier = 0;
+
+   const std::string kIdentifier   = "%f";
+   const size_t kIdentifierSize = 2;
+
+
+   std::string fractional_to_string(size_t fractional, size_t cutOffDivider) {
```
Comment (KjellKod) [inline src/time.cpp]: this has to be  moved to the .h/cpp so that it can be put under test

File: src/time.cpp [hunk 2]
```diff
@@ -2,17 +2,53 @@
...
 #include <cassert>
 #include <iomanip>
+#ifdef __MACH__
+#include <sys/time.h>
+#endif
+
+
+namespace {
+   const size_t kMsDigitsToCut = 1000000; // 3 digits left of 1000 000 000
+   const size_t kUsDigitsToCut = 1000; // 6 digits left
```
Comment (KjellKod) [inline src/time.cpp]: think about the naming + what is needed to be here and what can be defined static namespace or in the function?

File: src/time.cpp [hunk 3]
```diff
@@ -21,21 +57,21 @@ namespace g3 {
...
          return oss.str();
 #else    // LINUX
          const size_t size = 1024;
          char buffer[size]; // IMPORTANT: check now and then for when gcc will implement std::put_time.
          //                    ... also ... This is way more buffer space then we need
 
          auto success = std::strftime(buffer, size, c_time_format, tmb);
-         if (0 == success)
-         {
+         // TODO: incorrect code
```
Comment (KjellKod) [inline src/time.cpp]: remove erroneous comment

File: src/time.cpp [hunk 4]
```diff
@@ -65,11 +127,40 @@ namespace g3 {
...
    }
 
-   /// returns a std::string with content of time_t as localtime formatted by input format string
-   /// * format string must conform to std::put_time
-   /// This is similar to std::put_time(std::localtime(std::time_t*), time_format.c_str());
 
-   std::string localtime_formatted(const std::time_t &time_snapshot, const std::string &time_format) {
+
+   size_t fractional_cutoff(const std::string& format_buffer, size_t pos) {
+      char  ch  = (format_buffer.size() > pos + kIdentifierSize ? format_buffer.at(pos + kIdentifierSize) : '\0');
```
Comment (KjellKod) [inline src/time.cpp]: make testable

File: src/time.cpp [hunk 5]
```diff
@@ -65,11 +127,40 @@ namespace g3 {
...
+   std::string localtime_formatted(const timespec& time_snapshot, const std::string& time_format) {
+      auto format_buffer = time_format;  // copying format string to a separate buffer
+
+      // iterating through every "%f" instance in the format string
+      auto identifierExtraSize = 0;
+      for (size_t pos = 0; (pos = format_buffer.find(kIdentifier, pos)) != std::string::npos; pos += kIdentifierSize + identifierExtraSize) {
+         // figuring out whether this is nano, micro or milli identifier
+         auto cutoff = fractional_cutoff(format_buffer, pos);
+         auto value = fractional_to_string(time_snapshot.tv_nsec, cutoff);
+         auto identifierExtraSize = cutoff == kNsDefaultDigits ? kNoDigitIdentifier : kDigitIdentifier;
```
Comment (KjellKod) [inline src/time.cpp]: go through naming

File: src/time.cpp [hunk 6]
```diff
@@ -65,11 +127,40 @@ namespace g3 {
...
+
+   std::string localtime_formatted(const timespec& time_snapshot, const std::string& time_format) {
+      auto format_buffer = time_format;  // copying format string to a separate buffer
+
+      // iterating through every "%f" instance in the format string
+      auto identifierExtraSize = 0;
+      for (size_t pos = 0; (pos = format_buffer.find(kIdentifier, pos)) != std::string::npos; pos += kIdentifierSize + identifierExtraSize) {
+         // figuring out whether this is nano, micro or milli identifier
+         auto cutoff = fractional_cutoff(format_buffer, pos);
+         auto value = fractional_to_string(time_snapshot.tv_nsec, cutoff);
```
Comment (KjellKod) [inline src/time.cpp]: consider having the function which makes more sense
```g3::fractional_to_string(time_snapshot /*i.e. it takes the timespec*/,   FractionalType::us/ms/ns);```
it's just way more readable than the ```1000000000, 1000000, 1000``` and with one ```if-else``` switch all the nastiness is dealt with

File: src/time.cpp [hunk 7]
```diff
@@ -65,11 +127,40 @@ namespace g3 {
...
-   std::string localtime_formatted(const std::time_t &time_snapshot, const std::string &time_format) {
+
+   size_t fractional_cutoff(const std::string& format_buffer, size_t pos) {
+      char  ch  = (format_buffer.size() > pos + kIdentifierSize ? format_buffer.at(pos + kIdentifierSize) : '\0');
+      size_t cutOff = 0;
+      switch (ch) {
+         case '3': cutOff = kMsDigitsToCut; break;
+         case '6': cutOff = kUsDigitsToCut; break;
+         case '9': cutOff = kNsDigitsToCut; break;
+         default: cutOff = kNsDefaultDigits; break;
```
Comment (KjellKod) [inline src/time.cpp]: consider instead ```Fractional::ns/us/ms```

**Comment by gimesketvirtadieni** [comment]
Hi

I can see you merged milli/micro/nano sec functionality into a dedicated branch. Did I understand you correctly that you will make more refactoring of this functionality based on your comments above? Or do you want me to open one more pull request?
BTW there are few corrections required for your changes to make them compile on Linux, that is why Travis fails ;)

**Comment by KjellKod** [comment]
Thanks. No it's OK. I would have taken care of it already but I've been burning the midnight oil lately with work for my company

File: Build.cmake [hunk 8]
```diff
@@ -52,7 +52,7 @@ ELSEIF("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
        set(CMAKE_CXX_FLAGS "-Wall -Wunused -std=c++11  -pthread -D_GLIBCXX_USE_NANOSLEEP -D_GLIBCXX_USE_SCHED_YIELD")
    ELSE()
        set(PLATFORM_LINK_LIBRIES rt)
-       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -rdynamic -Wunused -std=c++11 -pthread -D_GLIBCXX_USE_NANOSLEEP -D_GLIBCXX_USE_SCHED_YIELD")
+       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -rdynamic -Wunused -std=c++11 -pthread -lrt -D_GLIBCXX_USE_NANOSLEEP -D_GLIBCXX_USE_SCHED_YIELD")
```
Comment (KjellKod) [inline Build.cmake]: is this change needed for clock_gettime?

File: scripts/buildAndRunTests.sh [hunk 9]
```diff
@@ -5,37 +5,49 @@ set -ev
 unzip -o 3rdParty/gtest/gtest-1.7.0.zip -d 3rdParty/gtest
 
 
-if [ "$CXX" = "g++" ]; then export CXX=g++-4.8; fi
-if [ "$CXX" = "clang++" ]; then export CXX=clang++-3.4; fi
-echo $TRAVIS_OS_NAME
-echo $CXX
+#if [ "$CXX" = "g++" ]; then export CXX=g++-5; fi
```
Comment (KjellKod) [inline scripts/buildAndRunTests.sh]: remove commented away code.

File: src/time.cpp (hunk 3 above)
Comment (KjellKod) [inline src/time.cpp]: consider replacing the assert to something that will tricker process exit even for production code. The `if 0 == success` can then be removed also.

File: test_unit/test_message.cpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#endif
+  }
+// This does not work. Other kinds of fatal exits (on Windows) seems to be used instead of exceptions
+// Maybe a signal handler catch would be better? --- TODO: Make it better, both failing and correct
+catch(...)
+{
+  ADD_FAILURE() << "On this platform the library does not support given (C++11?) specifiers";
+  return;
+}
+ASSERT_TRUE(true); // no exception. all good
```
Comment (KjellKod) [inline test_unit/test_message.cpp]: Add tests for the new functions that are not yet under test.

File: src/time.cpp (hunk 3 above)
Comment (KjellKod) [inline src/time.cpp]: No code change but added a comment so that code readers that have not seen this pattern before will be less baffled by it.    It is essentially a way of exiting with a clear message in debug mode. In production mode a safe value is instead returned.

File: src/time.cpp [hunk 11]
```diff
@@ -99,9 +99,12 @@ namespace g3 {
       ts->tv_sec  = now.tv_sec;
       ts->tv_nsec = now.tv_usec * 1000;
       return 0;
+#elif(defined(WIN32) || defined(_WIN32) || defined(__WIN32__))
+	   return timespec_get(ts, TIME_UTC);
```
Comment (KjellKod) [inline src/time.cpp]: This one was annoying m. std::timespec_get(...) doesn't exist but timespec_get(...) does

File: test_unit/test_concept_sink.cpp [hunk 12]
```diff
@@ -150,10 +150,10 @@ TEST(ConceptSink, DeletedSink__Exptect_badweak_ptr___exception) {
 }
 
 namespace {
-   typedef std::shared_ptr<std::atomic<bool >> AtomicBoolPtr;
-   typedef std::shared_ptr<std::atomic<int >> AtomicIntPtr;
-   typedef vector<AtomicBoolPtr> BoolList;
-   typedef vector<AtomicIntPtr> IntVector;
+   using AtomicBooleanPtr = std::shared_ptr<std::atomic<bool>>;
```
Comment (KjellKod) [inline test_unit/test_concept_sink.cpp]: Weird name lash with Windows

**Comment by KjellKod** [comment]
@gimesketvirtadieni PR updated for Windows/OSX/Linux.

Thanks for the work on this
